### PR TITLE
Support unbatched models in DALI backend

### PR DESCRIPTION
**Pull request by banasraf**
This PR enables unbatched models support in DALI backend and creates a special execution path for such models.

Unbatched model is a model with max_batch_size set to 0. In such case Triton does not interpret the first dimension of tensors as batch_size, which e.g. disables the dynamic batching. 

This mode is best fitting for our streamed video use-case because we in that case we always want to handle requests one-by-one.

This required changes in the config validation/autofill (config_tools/*) and separate, simplified ExecuteUnbatched method in DaliModelInstance.

Signed-off-by: Rafal <rbanas@nvidia.com>

### CODE REVIEW

File: src/dali_model_instance.cc [hunk 1]
```diff
@@ -149,37 +205,66 @@ InputsInfo DaliModelInstance::GenerateInputs(const std::vector<TritonRequest>& r
...
+  int num_samples = 0;
+  for (uint32_t input_idx = 0; input_idx < request.InputCount(); ++input_idx) {
+    auto input = request.InputByIdx(input_idx);
+    auto input_byte_size = input.ByteSize();
+    auto input_buffer_count = input.BufferCount();
+    auto meta = input.Meta();
+    auto &idescr = inputs[input_idx];
+    for (uint32_t buffer_idx = 0; buffer_idx < input_buffer_count; ++buffer_idx) {
+      auto buffer = input.GetBuffer(buffer_idx, device_type_t::CPU, GetDaliDeviceId());
+      idescr.buffers.push_back(buffer);
```
Comment (jantonguirao) [inline src/dali_model_instance.cc:239]: Don't know how heavy that object is, but maybe consider std::move ?

File: src/dali_model.h [hunk 2]
```diff
@@ -254,6 +259,10 @@ class DaliModel : public ::triton::backend::BackendModel {
 
   void ReadPipelineProperties() {
     int config_max_batch_size = ReadMaxBatchSize(model_config_);
+    if (config_max_batch_size == 0) {
```
Comment (jantonguirao) [inline src/dali_model.h:262]: and if it's 1? Is it still batched?

**Review by jantonguirao — APPROVED** [review]

File: src/utils/triton.h [hunk 3]
```diff
@@ -135,6 +135,10 @@ class TritonError : public UniqueHandle<TRITONSERVER_Error *, TritonError>, publ
     return TritonError(err);
   }
 
+  static TritonError Success() {
+    return TritonError(nullptr);
+  }
+
```
Comment (szalpal) [inline src/utils/triton.h:141]: I believe this looks fishy. I get, that `TRITONSERVER_Error == nullptr` denotes no error, but `TritonError(nullptr)` is still an `std::exception` object, which suggests an error.

File: src/dali_model_instance.h [hunk 4]
```diff
@@ -94,12 +99,17 @@ class DaliModelInstance : public ::triton::backend::BackendModelInstance {
   ProcessingMeta ProcessRequests(const std::vector<TritonRequest>& requests,
                                  const std::vector<TritonResponse>& responses);
 
+  TimeInterval ProcessRequest(const TritonRequest &request);
+
```
Comment (szalpal) [inline src/dali_model_instance.h:103]: Why is there no `responses` in the signature?

File: src/dali_model_instance.h [hunk 5]
```diff
@@ -55,11 +55,16 @@ class DaliModelInstance : public ::triton::backend::BackendModelInstance {
 
   void Execute(const std::vector<TritonRequest>& requests);
 
+  void ExecuteBatched(const std::vector<TritonRequest>& requests);
+
+  void ExecuteUnbatched(const std::vector<TritonRequest> &requests);
```
Comment (szalpal) [inline src/dali_model_instance.h:60]: I see that `Execute` handles both `ExecuteBatched` and `ExecuteUnbatched`. Do these functions have to be public?

File: src/dali_executor/io_descriptor.h [hunk 6]
```diff
@@ -56,11 +56,29 @@ template<typename T>
 struct IODescr {
   IOMeta meta{};
   std::vector<BufferDescr<T>> buffers{};
+
+  void append(const IODescr &other) {
+    if (meta.shape.num_samples() == 0) {
+      meta = other.meta;
```
Comment (szalpal) [inline src/dali_executor/io_descriptor.h:62]: Could you add documentation to this function?

File: qa/L0_unbatched_model/model_repository/model.dali/config.pbtxt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  {
+    name: "DALI_OUTPUT_0"
+  }
+]
+
+output [
+  {
+    name: "DALI_OUTPUT_1"
+  }
+]
```
Comment (szalpal) [inline qa/L0_unbatched_model/model_repository/model.dali/config.pbtxt:34]: Would this also work?
```
output [
  {
    name: "DALI_OUTPUT_0"
  },
  {
    name: "DALI_OUTPUT_1"
  }
]
```

File: src/dali_executor/dali_executor.h [hunk 8]
```diff
@@ -103,6 +107,7 @@ class DaliExecutor {
   ThreadPool thread_pool_;
   std::map<std::string, IOBuffer<CPU>> cpu_buffers_;
   std::map<std::string, IOBuffer<GPU>> gpu_buffers_;
+  bool input_consumed_;
```
Comment (szalpal) [inline src/dali_executor/dali_executor.h:110]: Could you document, what does that flag mean? Judging from the name, it suggests, that the `DaliExecutor` can operate only on one input.

File: src/dali_model.h (hunk 2 above)
Comment (szalpal) [reply src/dali_model.h:262]: I believe so. After all, `batch_size=1` seems like a valid configuration.

File: src/config_tools/config_tools.h [hunk 9]
```diff
@@ -106,28 +106,43 @@ void ValidateDtypeConfig(TritonJson::Value &io_object, const std::string &name,
 /**
  * @brief Auto-fills `config_io`'s dimensions field with value `model_io_shape`.
  * `config` must be a top-level TritonJson object containing `config_io`
+ *
+ * If `add_batch_dim` is set to true, the model_io_shape will be prepended with -1
+ * to extend it with the batch dimension.
```
Comment (szalpal) [inline src/config_tools/config_tools.h:111]: There's a reference to `add_batch_dim`, but there's no such argument. Did you mean `batched_model`?

File: src/dali_model_instance.h (hunk 4 above)
Comment (banasraf) [reply src/dali_model_instance.h:103]: ProcessRequest creates and send the responses - we don't know how many responses will be there for a request, so I cannot create them  before processing the request.

File: src/dali_executor/dali_executor.h (hunk 8 above)
Comment (banasraf) [reply src/dali_executor/dali_executor.h]: I documented the InputsConsumed method

File: qa/L0_unbatched_model/model_repository/model.dali/config.pbtxt (hunk 7 above)
Comment (banasraf) [reply qa/L0_unbatched_model/model_repository/model.dali/config.pbtxt]: works, changed

File: src/dali_executor/io_descriptor.h (hunk 6 above)
Comment (banasraf) [reply src/dali_executor/io_descriptor.h]: added

File: src/dali_model_instance.h (hunk 5 above)
Comment (banasraf) [reply src/dali_model_instance.h]: done

File: src/utils/triton.h (hunk 3 above)
Comment (banasraf) [reply src/utils/triton.h]: True. I removed this Success method. Instead I changed the SendResponse to not require providing the error if it's supposed to be success (nullptr)

File: src/config_tools/config_tools.h (hunk 9 above)
Comment (banasraf) [reply src/config_tools/config_tools.h:111]: fixed

File: src/dali_model_instance.cc (hunk 1 above)
Comment (banasraf) [reply src/dali_model_instance.cc]: Good point. I added move constructors to some of the classes and minimized copying

**Review by szalpal — APPROVED** [review]

**Pull request merged**